### PR TITLE
First version of algorithm without optional attendees

### PR DESCRIPTION
**Pull request by jeinygroove**

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,10 +14,61 @@
...
-import java.util.Collection;
+import java.util.*;
 
 public final class FindMeetingQuery {
+
+  /**
+   * Returns collection of time slots, in which we can appoint the meeting.
+   * @param events   All events in the day.
+   * @param request  Information about event that we want to add (attendees, required duration).
+   * */
```
Comment (jgaillard85) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: fix formatting, the last line of a javadoc is simply */

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,61 @@
...
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // hash map, where key is the time moment and object is the number of attendees from the request,
+    // who will be busy or free from that time moment
+    HashMap<Integer, Long> timePoints = new HashMap<>();
+    Collection<String> attendees = request.getAttendees();
+
+    long requiredAmount = attendees.size();
```
Comment (jgaillard85) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: variable name could be more illuminating, maybe something like numberOfAttendeesRequired

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,61 @@
...
+
+        timePoints.put(startTime, timePoints.getOrDefault(startTime, 0L) - eventAttendees.size());
+        timePoints.put(endTime, timePoints.getOrDefault(endTime, 0L) + eventAttendees.size());
+      }
+    }
+
+    List<Integer> sortedTimes = new ArrayList<>(timePoints.keySet());
+    Collections.sort(sortedTimes);
+
+    long currentAmount = 0;
```
Comment (jgaillard85) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Maybe here call it something like numberOfAttendeesCurrentlyAvailable

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,61 @@
...
+
+  /**
+   * Returns collection of time slots, in which we can appoint the meeting.
+   * @param events   All events in the day.
+   * @param request  Information about event that we want to add (attendees, required duration).
+   * */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // hash map, where key is the time moment and object is the number of attendees from the request,
+    // who will be busy or free from that time moment
```
Comment (jgaillard85) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Not very clear, explain that it is the overall change in the number of available attendees at that particular time.
Maybe a better name for the variable could help convey that better as well.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,64 @@
 
 package com.google.sps;
 
-import java.util.Collection;
+import java.util.*;
 
 public final class FindMeetingQuery {
+
+  /**
+   * Returns collection of time slots, in which we can appoint the meeting.
```
Comment (amaurymedeiros) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: "book" the meeting

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,64 @@
...
+   * @param events   All events in the day.
+   * @param request  Information about event that we want to add (attendees, required duration).
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // hash map, where key is the moment of time and object is
+    // the change in the number of available attendees at that particular time
```
Comment (amaurymedeiros) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I think for "object" you mean "value". From the documentation, it's not too clear what the value represents. A suggestion is to include an example in the comment to make it better for the reader to understand

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -14,10 +14,64 @@
...
+
+    // add start and end times of the day
+    timePointsOfAvailabilityChange.put(0, numberOfAttendeesRequired);
+    timePointsOfAvailabilityChange.put(TimeRange.WHOLE_DAY.end(), -numberOfAttendeesRequired);
+
+    // add all time point, where the number of free attendees is changing
+    for (Event event : events) {
+      // the event can have attendees that are not present in the request,
+      // so we shouldn't count them
+      Collection<String> eventAttendees = new ArrayList<>(event.getAttendees());
```
Comment (amaurymedeiros) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: if I understood correctly, these are the attendees that are not on the request. So two things here:

- change the inline comment slightly to say "the event can _only_ have ...."
- change the variable name to something more precise, e.g. `eventAttendeesNotOnMeetingRequest`

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -14,10 +14,64 @@
...
+
+        timePointsOfAvailabilityChange.put(startTime,
+                timePointsOfAvailabilityChange.getOrDefault(startTime, 0L) - eventAttendees.size());
+        timePointsOfAvailabilityChange.put(endTime,
+                timePointsOfAvailabilityChange.getOrDefault(endTime, 0L) + eventAttendees.size());
+      }
+    }
+
+    List<Integer> sortedTimes = new ArrayList<>(timePointsOfAvailabilityChange.keySet());
+    Collections.sort(sortedTimes);
```
Comment (amaurymedeiros) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit/optional: maybe it's more readable if you use Streams.

`List<Integer> sortedTimes = timePointsOfAvailabilityChange.keySet().stream().sorted().collect(Collectors.toList())`

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 9]
```diff
@@ -14,10 +14,64 @@
...
+      Collection<String> eventAttendees = new ArrayList<>(event.getAttendees());
+      eventAttendees.retainAll(attendees);
+
+      // if intersection isn't empty
+      if (!eventAttendees.isEmpty()) {
+        int startTime = event.getWhen().start();
+        int endTime = event.getWhen().end();
+
+        timePointsOfAvailabilityChange.put(startTime,
+                timePointsOfAvailabilityChange.getOrDefault(startTime, 0L) - eventAttendees.size());
```
Comment (amaurymedeiros) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: you can extract `eventAttendees.size()` to a variable as well (e.g. `numAttendees`)

**Review by jgaillard85 — APPROVED** [review]

**Pull request merged**